huff0/_generate/gen.go

```diff
@@ -0,0 +1,366 @@
+package main
+
+//go:generate go run gen.go -out decompress_amd64.s -stubs delme.go -pkg=huff0
+
+import (
+	"flag"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	_ "github.com/klauspost/compress"
+
+	. "github.com/mmcloughlin/avo/build"
+	"github.com/mmcloughlin/avo/buildtags"
+	"github.com/mmcloughlin/avo/gotypes"
+	. "github.com/mmcloughlin/avo/operand"
+	"github.com/mmcloughlin/avo/reg"
+)
+
+func main() {
+	flag.Parse()
+	out := flag.Lookup("out")
+	os.Remove(filepath.Join("..", out.Value.String()))
+	stub := flag.Lookup("stubs")
+	if stub.Value.String() != "" {
+		os.Remove(stub.Value.String())
+		defer os.Remove(stub.Value.String())
+	}
+
+	Constraint(buildtags.Not("appengine").ToConstraint())
+	Constraint(buildtags.Not("noasm").ToConstraint())
+	Constraint(buildtags.Term("gc").ToConstraint())
+	Constraint(buildtags.Not("noasm").ToConstraint())
+
+	decompress := decompress4x{}
+	decompress.generateProcedure("decompress4x_main_loop_amd64")
+	decompress.bmi2 = true
+	decompress.generateProcedure("decompress4x_main_loop_bmi2")
+
+	decompress8b := decompress4x8bit{}
+	decompress8b.generateProcedure("decompress4x_8b_main_loop_amd64")
+	decompress8b.bmi2 = true
+	decompress8b.generateProcedure("decompress4x_8b_main_loop_bmi2")
+
+	Generate()
+	b, err := ioutil.ReadFile(out.Value.String())
+	if err != nil {
+		panic(err)
+	}
+	const readOnly = 0444
+	err = ioutil.WriteFile(filepath.Join("..", out.Value.String()), b, readOnly)
+	if err != nil {
+		panic(err)
+	}
+	os.Remove(out.Value.String())
```

Comment:
This is bizarre too. You're copying it and rewriting it with readonly permissions? Why not just do this outside the avo generator, with a `chmod` call?

Reply:
Removed in #569